python/ray/tune/pbt.py

```diff
@@ -0,0 +1,174 @@
+                else:
+                    config_dictionary = {param: self._round_up_to_even(random.choice([0.8, 1.2]) * config_dictionary[param]) for param in config_dictionary if param != "env"}
+                config_dictionary["env"] = best_trial.config["env"]
+                print("new hyperparamter configuration: {0}".format(config_dictionary))
+                checkpoint = best_trial.checkpoint()
+                print(checkpoint)
+                new_trial = Trial(
+                    trainable_name = best_trial.trainable_name,
+                    config = config_dictionary,
+                    local_dir= best_trial.local_dir,
+                    experiment_tag= best_trial.experiment_tag + mutate_string,
+                    resources = best_trial.resources,
+                    stopping_criterion= best_trial.stopping_criterion,
+                    checkpoint_freq= best_trial.checkpoint_freq,
+                    restore_path= checkpoint,
+                    upload_dir= best_trial.upload_dir
+                )
+                
+                #mutate the hyperparamters the new trial, last index of config is always the "env"
+                #so exclude that, temporarily we "perturb" each hyperparameter by a factor 1.2 or 0.8 and then round it to the nearest
+                #even integer
+                print(new_trial.config)
+                new_trial.experiment_tag = trial.experiment_tag + "-mutated@" + str(time)
+                #add the new trial to the trialrunner, and then stop the
+                #  trial it replaced
+                trial_runner._launch_trial(new_trial)
+                trial_runner.add_trial(new_trial)
+                self._results[new_trial] = self._results[trial]
+                self._time_tracker[new_trial] = self._time_tracker[trial]
+                self._results.pop(trial)
+                self._time_tracker.pop(trial)
+                best_trial.stop() if best_trial_was_stopped else print("best trial continues")
+                return TrialScheduler.STOP
+                
+
+
+        else:
+            print("not ready")
+        
+        return TrialScheduler.CONTINUE
+
+    def on_trial_complete(self, trial_runner, trial, result):
+        self._results[trial].append(result)
+        self._completed_trials.add(trial)
+        print("finished a trial")
+        min_time = 0
+        best_trial = None
+        for trial in self._completed_trials:
+            last_result = self._results[trial][-1]
+            if getattr(last_result, self._time_attr) < min_time or min_time == 0:
+                min_time = getattr(last_result, self._time_attr)
+                best_trial = trial
+        print("The Best Trial is currently {0} finishing in {1} iterations, with the hyperparameters of {2}".format(best_trial, min_time, best_trial.config))
+
+    def _default_ready(self, result, trial, time, time_diff = 4):
```

Comment:
Shall we make `time_diff` a constructor param? Maybe it can be the same as grace_period, and you can call it `perturbation_interval`.